p2p/src/sync/mod.rs

```diff
@@ -502,10 +574,40 @@ where
         }
     }
 
+    async fn process_error(
+        &mut self,
+        peer_id: T::PeerId,
+        request_id: T::RequestId,
+        error: net::RequestResponseError,
+    ) -> error::Result<()> {
+        match error {
+            net::RequestResponseError::ConnectionClosed => {
+                self.unregister_peer(peer_id);
+            }
+            net::RequestResponseError::Timeout => {
+                if let Some(request) = self.requests.remove(&request_id) {
+                    log::warn!("outbound request timed out, resend");
+
+                    match request.request_type {
+                        RequestType::GetHeaders => {
+                            let locator =
+                                self.consensus_handle.call(|this| this.get_locator()).await??;
+                            self.send_header_request(peer_id, locator).await?;
+                        }
+                        RequestType::GetBlocks(block_id) => {
+                            self.send_block_request(peer_id, block_id).await?;
+                        }
+                    }
+                }
```

Comment:
Is there a limit on the number of resend attempts?

Reply:
No but I think there should be. Maybe 3?

Reply:
Usually, if communication fails, the connection is just dropped.

Reply:
I don't think one failed request automatically means a failed communication. We've also specified that only after three failed PING/PONG retries is the remote peer considered unresponsive. Especially for public nodes during peak traffic there will certainly come a time when it must drop an incoming message because its own queues are full, meaning the request will timeout. But IMO that doesn't mean that after the first dropped message the communication has failed, just that they were not able to service us at that time and we should retry. But I agree with Lukas that there should be a limit for how many times we retry before closing the connection.